Supporting -std=c++17 flag

AntonBikineev opened the pull request.

Naios commented on CMakeLists.txt:
```diff
@@ -959,42 +959,58 @@ if(NOT MSVC)
-        add_definitions(-DBOOST_NO_CXX14_CONSTEXPR)
-      endif()
+    if(HPX_WITH_CXX17)
+      set(CXX_FLAG -std=c++17)
```
Since Visual Studio 2015 the Microsoft compiler require a switch for later standard versions too `/std:c++14` and ` /std:c++latest` (see [here](https://blogs.msdn.microsoft.com/vcblog/2016/06/07/standards-version-switches-in-the-compiler/)).

A (maybe better) build system driven way of depending on compiler features is the cmake [`target_compile_features`](https://cmake.org/cmake/help/v3.8/manual/cmake-compile-features.7.html) property, however this feature was introduced in cmake version 3.1.

AntonBikineev replied:
Naios, thanks for the note. Yeah, target_compile_features would be a great independent way of handling these cmd args once we introduce support for cmake-3.1.

hkaiser replied:
We've raised the minimal version of cmake to 3.0.2 already, I wouldn't mind making 3.1 the minimal requirement.

hkaiser replied:
Well, since your test for `HPX_HAVE_CXX11_STD_SHUFFLE` includes `std::random_device` we should be fine as it is.

Naios commented on src/util/init_ini_data.cpp:
```diff
@@ -451,8 +452,14 @@ namespace hpx { namespace util
+        std::random_device random_device;
+        std::mt19937 generator(random_device());
+        std::shuffle(libdata.begin(), libdata.end(), std::move(generator));
+#else
```
Wouldn't it be better if we provide a generic `hpx::container_shuffle` for encapsuling the best backend which is currently available, instead of replacing all usage cases of `std::random_shuffle` with conditional compilation?

AntonBikineev replied:
It's the only place where random-shuffle is used. That said I think that minimal versions of compilers that we support are good enough to have std shuffle in their default standard libraries. So we probably don't even need this cmake test at all. What do others think about it?

hkaiser replied:
In my experience, `std::random_device` is a bigger hurdle to compatibility... It's one of the reasons we normally use the boost counterpart in other places. But I have not checked recently whether the situation has changed.

AntonBikineev replied:
@hkaiser do you suggest to use boost::random_device here?

hkaiser replied:
Well, since your test for  `HPX_HAVE_CXX11_STD_SHUFFLE`  includes  `std::random_device`  we should be fine as it is.

hkaiser commented on CMakeLists.txt:
```diff
@@ -959,42 +959,58 @@ if(NOT MSVC)
+            endif()
           endif()
         endif()
       endif()
```
I'd suggest moving the whole `-std=XXX` business into a separate cmake file to avoid cluttering the main one.

hkaiser commented:
@AntonBikineev I have changed this PR since it was approved. Could you have another look, please?

AntonBikineev commented:
@hkaiser it looks great. Sorry, I just postponed addressing your comments. I also remember there was an example which failed to compile with -std=c++1z (it used Boost.Phoenix under Boost.Spirit and there was std::random_shuffle or something in there...)
